code/client/src/components/AddressInput.jsx

```diff
@@ -0,0 +1,103 @@
+import { SearchOutlined } from '@ant-design/icons'
+import { Select } from 'antd'
+import React, { useCallback } from 'react'
+import { useDispatch, useSelector } from 'react-redux'
+import walletActions from '../state/modules/wallet/actions'
+import util, { useWindowDimensions } from '../util'
+
+/**
+ * Renders address input that provides type ahead search for any known ONE addresses.
+ * Known addresses are addresses that have been entered by user for at least once.
+ * If an address is entered in none one1... format, the address will be converted and saved as one1... format.
+ */
+const AddressInput = ({ setAddressCallback, currentWallet, addressValue, extraSelectOptions, knownAddressKey }) => {
+  const dispatch = useDispatch()
+
+  const wallets = useSelector(state => Object.keys(state.wallet.wallets).map((k) => state.wallet.wallets[k]))
+
+  const knownAddresses = useSelector(state => state.wallet.knownAddresses)
+
+  const network = useSelector(state => state.wallet.network)
+
+  const { isMobile } = useWindowDimensions()
+
+  const onChangeAddress = useCallback((address) => {
+    setAddressCallback(address)
+  }, [setAddressCallback])
+
+  /**
+   * Determines if the input wallet wallet is for the current wallet. Applicable for existing wallet management.
+   */
+  const notCurrentWallet = useCallback((inputWalletAddress) =>
+    !currentWallet || currentWallet?.address !== inputWalletAddress,
+  [currentWallet])
+
+  /**
+   * This is interim wallet object that uses known wallet addresses to be used to render the address selection with
+   * existing wallet object. The address key is unique as knownOneAddress. Note that this will not be saved as actual wallet.
+   */
+  const knownWallets = (knownAddresses ? knownAddresses[knownAddressKey] || [] : []).map((knownOneAddress) => ({
+    network,
+    knownOneAddress
+  }))
+
+  const addressOptions = [...wallets, ...knownWallets]
+
+  const existingKnownAddress = useCallback((oneAddress) => {
+    return addressOptions.find((wallet) =>
+      util.safeOneAddress(wallet.address) === oneAddress || wallet.knownOneAddress === oneAddress)
+  }, [addressOptions])
+
+  const onSelectAddress = useCallback((oneAddress) => {
+    const validOneAddress = util.safeOneAddress(oneAddress)
+
+    if (validOneAddress && !existingKnownAddress(validOneAddress)) {
+      dispatch(walletActions.addKnownAddress({ knownAddressKey, oneAddress: validOneAddress }))
```

Comment:
I think it would be beneficial to store the addresses as key-value format, with key equal to the normalized address, and value equal to an object consisting of: label (e.g. wallet name), network (i.e. currently selected network), creationTime (i.e. when it is added), lastUsedTime, and numUsed (e.g. increment by one each time it is selected). The reason is that we will need to allow users to re-label the address later, and later they would expect us to intelligently rank the addresses based on how frequently they use each address, or how old they are. There might be more extensions in the future. Storing them as objects would allow us easily add more extensions, whereas array would not.

Reply:
network is quite important - it must be stored, to allow the options automatically filter by network. We don't want the users accidentally send to an address on a different network - the money would have been lost

Reply:
let's at  least support deletion - otherwise the list of addresses could add up and become overwhelming quite quickly

Reply:
Address changed to object in state.